### PR TITLE
Feature: Expansion of Preferences Manager

### PR DESCRIPTION
**Pull request by AsimRibo**
## Summary

To make finding specific preferences easier, I decided to split our logic for displaying preferences into 2 parts: all preferences and targeted ones. 
## Changes
`PreferencesTab` is now split into 2 parts, tab itself contains `TargetedPreferences` and there is a button that leads to `AllPreferences`. I added additional collector that will collect only files and properties sent by clients via `watch` function.

### Recycler
Since a lot data was being collected, Sentinel started stuttering and crashing sometimes. I switched item displaying to recyclerView. Adapter has been a bit harder to write since we had a list of data that also contains a list of data, and everything needs to be displayed. SOLUTION: I went with flattening the preferences data and 2 view holders, where one is preference file item and one is preference property (or child as I called it)

### Type

- [x] **Feature**: This pull request introduces a new feature.
- [ ] **Bug fix**: This pull request fixes a bug.
- [ ] **Refactor**: This pull request refactors existing code.
- [ ] **Documentation**: This pull request updates documentation.
- [ ] **Other**: This pull request makes other changes.

#### Additional information

- [ ] This pull request introduces a **breaking change**.

## Checklist

- [x] I have performed a self-review of my own code.
- [x] I have tested my changes, including edge cases.
- [ ] I have added necessary tests for the changes introduced (if applicable).
- [ ] I have updated the documentation to reflect my changes (if applicable).

## Additional notes
I will upload documentation a bit later.

Currently if you update an item in `AllPreferences` and go back to `TargetedPreferences,` that change won't be shown there. I will try to fix it when I find time, but I don't think it is a blocker for now.

Also, since we have many files, if you minimize a couple of files and they are still visible on the screen when you click to change a preference property and save it, all other files will expand and you will be scrolled away from that specific preference[if minimized items are not visible, this won't happen]. This is because when preference property is changed, we reload entire screen. We would have to adjust logic so we don't reload entire screen but this can be done later.

### CODE REVIEW

**Comment by AsimRibo** [comment]
We should probably disable this lexicographic ordering, not sure about benefits?

File: sentinel/src/main/kotlin/com/infinum/sentinel/ui/main/preferences/shared/model/PreferencesItem.kt [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+internal fun List<PreferencesData>.flatten(): List<PreferencesItem> = flatMap { data ->
+    listOf(PreferencesItem.Parent(name = data.name, isExpanded = data.isExpanded)) +
+        if (data.isExpanded) {
+            data.values.map { (preferenceType, label, value) ->
+                PreferencesItem.Child(preferenceType, label, value, parentName = data.name)
+            }
+        } else {
+            emptyList()
+        }
```
Comment (KCeh) [inline sentinel/src/main/kotlin/com/infinum/sentinel/ui/main/preferences/shared/model/PreferencesItem.kt:29]: 👌 
Interesting workaround

**Review by KCeh — COMMENTED** [review]
Looks good to me 🚀 
I like idea of "parent" and "child" in recycler viewer to simplify logic.

I just see potential area where we can improve:
Unify targeted and all preferences logic and presentation. 
To me, it seems that both logic and presentations are same(ish). For targeted prefs we just pass "filter". And for all prefs we pass empty "filter"/all included filter (depending on how logic works out)
Could we use just one fragment and one VM for both things? If needed we can add nav argument to make distinction which screen is in question.
  

Correct me if I am wrong. And if you see potential issues.

**Comment by KCeh** [comment]
> We should probably disable this lexicographic ordering, not sure about benefits?

What is the source of this "problem"
Ktlint or detekt? We can update rules. Hopefully it will resolve things (and TBT I think just detekt should be enough)

File: sentinel/src/main/kotlin/com/infinum/sentinel/Sentinel.kt [hunk 2]
```diff
@@ -13,10 +13,30 @@ import java.util.Locale
...
+     *
+     * // This will contain default Sentinel tools + Chucker
+     * Sentinel.watch(setOf(ChuckerTool()))
+     *
+     * // This will contain default Sentinel tools + Chucker and all preference properties of MY_PREFS_FILE
+     * Sentinel.watch(setOf(ChuckerTool()), mapOf("MY_PREFS_FILE" to emptyList()))
+     *
+     * // This will contain default Sentinel tools + Chucker and only MY_PROPERTY property of MY_PREFS_FILE
+     * Sentinel.watch(setOf(ChuckerTool()), mapOf("MY_PREFS_FILE" to listOf("MY_PROPERTY")))
+     */
```
[the remark this answers is not in the document]
Comment (AsimRibo) [reply sentinel/src/main/kotlin/com/infinum/sentinel/Sentinel.kt:35]: > I know this might counter your design goal (selected prefs and all prefs on separate screens). But, can you watch all preferences in this selected preferences quick view? How? Regardless of answer could we add an example in docs to watch 2 separate files?

In selected preferences quick view, you can monitor multiple files but you have to provide names of those files. If that's what you are asking. IMO, it makes sense since app will have dozens of not so relevant files which are then cluttering your UI [but you can access them in AllFiles].

> How does the iOS version behave in the default state where a user of lib doesn't specify which preferences they want to watch

To my knowledge, they don't have option to display all files, they just display what clients want. They also have some logic of mapping where clients can provide readable names of properties (instead of KEY_ONE they can send mapping "key one" which is more readable lets say.) I will have to double check with them just in case.

File: sentinel/src/main/kotlin/com/infinum/sentinel/data/sources/raw/collectors/TargetedPreferencesCollector.kt [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+                        is Float -> Triple(PreferenceType.FLOAT, key, value)
+                        is Int -> Triple(PreferenceType.INT, key, value)
+                        is Long -> Triple(PreferenceType.LONG, key, value)
+                        is String -> Triple(PreferenceType.STRING, key, value)
+                        is Set<*> -> Triple(PreferenceType.SET, key, value as Set<String>)
+                        else -> null
+                    }
+                }
+
+            if (tuples.isNotEmpty()) PreferencesData(fileName, tuples) else null
```
[the remark this answers is not in the document]
Comment (AsimRibo) [reply sentinel/src/main/kotlin/com/infinum/sentinel/data/sources/raw/collectors/TargetedPreferencesCollector.kt:37]: I think it can be done. I would just have to add additional filtering of files since there we also fetch all preference files. 👍

**Comment by AsimRibo** [comment]
> Could we use just one fragment and one VM for both things? If needed we can add nav argument to make distinction which screen is in question.

I was being a bit lazy 😄 , but this can be done yes. We will just need some feature flag to toggle visible views since we have that all preferences button, and I plan on adding search functionality limited to `AllPreferences` (there is no space on TargetedPreferences IMO, maybe if I added collapsing toolbar but we will see).

**Comment by AsimRibo** [comment]
> What is the source of this "problem"
Ktlint or detekt? We can update rules. Hopefully it will resolve things (and TBT I think just detekt should be enough)

I think for lexicographic ordering it was detekt. We should probably update rules.

**Comment by KCeh** [comment]
> > What is the source of this "problem"
> > Ktlint or detekt? We can update rules. Hopefully it will resolve things (and TBT I think just detekt should be enough)
> 
> I think for lexicographic ordering it was detekt. We should probably update rules.

Yea sure, I will open task for it

File: sentinel/src/main/kotlin/com/infinum/sentinel/data/sources/raw/collectors/TargetedPreferencesCollector.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (KCeh) [reply sentinel/src/main/kotlin/com/infinum/sentinel/data/sources/raw/collectors/TargetedPreferencesCollector.kt:37]: I rechecked this code. 
So we want to have option to have both all preferences and only targeted ones?
Then I don't think we "unify" collectors easily. 
We would need to have invoke with arguments (filter list or flag at least). Because we would need to invoke the same code in 2 different places with 2 different behaviors.
I will try to define one more interface so we can have one implementation to cover both cases (right now I only see 2 separate implementations as implement by you as possible solution)

**Comment by KCeh** [comment]
I will rereview and retest this feature when I find the time

**Comment by KCeh** [comment]
> > What is the source of this "problem"
> > Ktlint or detekt? We can update rules. Hopefully it will resolve things (and TBT I think just detekt should be enough)
> 
> I think for lexicographic ordering it was detekt. We should probably update rules.

I disabled that rule 🥳

**Comment by AsimRibo** [comment]
I want to test this once more from my side and then I will merge.
